cpp/ql/src/experimental/Security/CWE/CWE-675/DoubleRelease.ql

```diff
@@ -0,0 +1,211 @@
+  CallMayNotReturn() {
+    // call that is known to not return
+    not exists(this.(ControlFlowNode).getASuccessor())
+    or
+    // call to another function that may not return
+    exists(CallMayNotReturn exit | getTarget() = exit.getEnclosingFunction())
+    or
+    exists(ThrowExpr tex | tex = this.(ControlFlowNode).getASuccessor())
+  }
+}
+
+/** Holds if there are no assignment expressions to the function argument. */
+pragma[inline]
+predicate checkChangeVariable(FunctionCall fc0, FunctionCall fc1, FunctionCall fc2) {
+  not exists(Expr exptmp |
+    (
+      exptmp = fc0.getArgument(0).(VariableAccess).getTarget().getAnAssignedValue() or
+      exptmp.(AddressOfExpr).getOperand() =
+        fc0.getArgument(0).(VariableAccess).getTarget().getAnAccess()
+    ) and
+    exptmp = fc1.getASuccessor*() and
+    exptmp = fc2.getAPredecessor*()
+  ) and
+  (
+    (
+      not fc0.getArgument(0) instanceof PointerFieldAccess and
+      not fc0.getArgument(0) instanceof ValueFieldAccess
+      or
+      fc0.getArgument(0).(VariableAccess).getQualifier() instanceof ThisExpr
+    )
+    or
+    not exists(Expr exptmp |
+      (
+        exptmp =
+          fc0.getArgument(0)
+              .(VariableAccess)
+              .getQualifier()
+              .(VariableAccess)
+              .getTarget()
+              .getAnAssignedValue() or
+        exptmp.(AddressOfExpr).getOperand() =
+          fc0.getArgument(0)
+              .(VariableAccess)
+              .getQualifier()
+              .(VariableAccess)
+              .getTarget()
+              .getAnAccess()
+      ) and
+      exptmp = fc1.getASuccessor*() and
+      exptmp = fc2.getAPredecessor*()
+    )
+  )
+}
+
+/** Holds if the underlying expression is a call to the close function. Provided that the function parameter does not change after the call. */
+predicate closeReturn(FunctionCall fc) {
+  fcloseCall(fc, _) and
+  not exists(Expr exptmp |
+    (
+      exptmp = fc.getArgument(0).(VariableAccess).getTarget().getAnAssignedValue() or
+      exptmp.(AddressOfExpr).getOperand() =
+        fc.getArgument(0).(VariableAccess).getTarget().getAnAccess()
+    ) and
+    exptmp = fc.getASuccessor*()
+  ) and
+  (
+    (
+      not fc.getArgument(0) instanceof PointerFieldAccess and
+      not fc.getArgument(0) instanceof ValueFieldAccess
+      or
+      fc.getArgument(0).(VariableAccess).getQualifier() instanceof ThisExpr
+    )
+    or
+    not exists(Expr exptmp |
+      (
+        exptmp =
+          fc.getArgument(0)
+              .(VariableAccess)
+              .getQualifier()
+              .(VariableAccess)
+              .getTarget()
+              .getAnAssignedValue() or
+        exptmp.(AddressOfExpr).getOperand() =
+          fc.getArgument(0)
+              .(VariableAccess)
+              .getQualifier()
+              .(VariableAccess)
+              .getTarget()
+              .getAnAccess()
+      ) and
+      exptmp = fc.getASuccessor*()
+    )
+  )
```

Comment:
This predicate, and `closeWithoutChangeBefore` below it, are large chunks of complicated code without any comments inside them explaining how they work.  This is difficult to review and will be difficult to maintain.  Would you be able to add comments explaining what each block of code does, and/or simplify the logic?

Reply:
these snippets are the `checkChangeVariable` predicate without one line. unfortunately I was not able to use it.
it would be great if you could tell me a typical approach for such situations. (general predicate `checkChangeVariable`)

Reply:
Ah, I see what's going on now.  If you change the type of `checkChangeVariable` to:
```
pragma[inline]
predicate checkChangeVariable(FunctionCall fc0, ControlFlowNode fc1, ControlFlowNode fc2)
```
I believe you can replace the body of this function with:
```
  fcloseCall(fc, _) and
  checkChangeVariable(fc, fc, fc.getEnclosingFunction())
```
and `closeWithoutChangeBefore` with:
```
  fcloseCall(fc, _) and
  checkChangeVariable(fc, fc.getEnclosingFunction().getEntryPoint(), fc)
```

Reply:
thanks for this idea.
I will run tests and will definitely use it.